tests/components/sharkiq/test_shark_iq.py

```diff
@@ -0,0 +1,283 @@
+
+from homeassistant.components.sharkiq import SharkIqUpdateCoordinator
+from homeassistant.components.sharkiq.vacuum import (
+    ATTR_ERROR_CODE,
+    ATTR_ERROR_MSG,
+    ATTR_LOW_LIGHT,
+    ATTR_RECHARGE_RESUME,
+    STATE_RECHARGING_TO_RESUME,
+    SharkVacuumEntity,
+)
+from homeassistant.components.vacuum import (
+    STATE_CLEANING,
+    STATE_DOCKED,
+    STATE_IDLE,
+    STATE_PAUSED,
+    STATE_RETURNING,
+    SUPPORT_BATTERY,
+    SUPPORT_FAN_SPEED,
+    SUPPORT_LOCATE,
+    SUPPORT_PAUSE,
+    SUPPORT_RETURN_HOME,
+    SUPPORT_START,
+    SUPPORT_STATE,
+    SUPPORT_STATUS,
+    SUPPORT_STOP,
+)
+from homeassistant.config_entries import ConfigEntriesFlowManager, ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import UpdateFailed
+
+from .const import (
+    SHARK_DEVICE_DICT,
+    SHARK_METADATA_DICT,
+    SHARK_PROPERTIES_DICT,
+    TEST_PASSWORD,
+    TEST_USERNAME,
+)
+
+from tests.async_mock import MagicMock, patch
+
+try:
+    import ujson as json
+except ImportError:
+    import json
+
+
+MockAyla = MagicMock(spec=AylaApi)  # pylint: disable=invalid-name
+
+
+def _set_property(self, property_name, value):
+    """Set a property locally without hitting the API."""
+    if isinstance(property_name, enum.Enum):
+        property_name = property_name.value
+    if isinstance(value, enum.Enum):
+        value = value.value
+    self.properties_full[property_name]["value"] = value
+
+
+async def _async_set_property(self, property_name, value):
+    """Set a property locally without hitting the API."""
+    _set_property(self, property_name, value)
+
+
+def _get_mock_shark_vac(ayla_api: AylaApi) -> SharkIqVacuum:
+    """Create a crude sharkiq vacuum with mocked properties."""
+    shark = SharkIqVacuum(ayla_api, SHARK_DEVICE_DICT)
+    shark.properties_full = deepcopy(SHARK_PROPERTIES_DICT)
+    return shark
+
+
+async def _async_list_devices(_) -> List[Dict]:
+    """Generate a dummy of async_list_devices output."""
+    return [SHARK_DEVICE_DICT]
+
+
+@patch.object(SharkIqVacuum, "set_property_value", new=_set_property)
+@patch.object(SharkIqVacuum, "async_set_property_value", new=_async_set_property)
+async def test_shark_operation_modes(hass: HomeAssistant) -> None:
+    """Test all of the shark vacuum operation modes."""
+    ayla_api = MockAyla()
+    shark_vac = _get_mock_shark_vac(ayla_api)
+    coordinator = SharkIqUpdateCoordinator(hass, None, ayla_api, [shark_vac])
+    shark = SharkVacuumEntity(shark_vac, coordinator)
+
+    # These come from the setup
+    assert isinstance(shark.is_docked, bool) and not shark.is_docked
+    assert (
+        isinstance(shark.recharging_to_resume, bool) and not shark.recharging_to_resume
+    )
+    # Go through the operation modes while it's "off the dock"
+    await shark.async_start()
+    assert shark.operating_mode == shark.state == STATE_CLEANING
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6090606)